hw/xfree86/common/xf86DGA.c

```diff
@@ -1313,28 +1313,23 @@ static void
 DGAClientStateChange(CallbackListPtr *pcbl, void *nulldata, void *calldata)
 {
     NewClientInfoRec *pci = (NewClientInfoRec *) calldata;
-    ClientPtr client = NULL;
     int i;
 
     for (i = 0; i < screenInfo.numScreens; i++) {
-        if (DGA_GETCLIENT(i) == pci->client) {
-            client = pci->client;
-            break;
-        }
-    }
+        if (pci->client && (DGA_GETCLIENT(i) == pci->client) &&
+            ((pci->client->clientState == ClientStateGone) ||
+             (pci->client->clientState == ClientStateRetained)))
```

Comment:
We probably should stop iterating if `DGA_GETCLIENT(i) == pci->client` is true, but the `clientState` check failed.

Reply:
yes, that brings a little bit better performance.
rewritten it to nested if's and a final break.